text/0000-generalized-arity-tuples.md

```diff
@@ -0,0 +1,277 @@
+
+```rust
+trait Last {
+    type Last;
+
+    fn last(self) -> Self::Last;
+}
+
+impl<ELEM> Last for (ELEM,) {
+    type Last = ELEM;
+
+    fn last(self) -> Self::Last {
+        self.elem
+    }
+}
+
+impl<ELEM, TAIL: Last> Last for Tuple<ELEM, TAIL> {
+    type Last = TAIL::Last;
+
+    fn last(self) -> Self::Last {
+        self.tail.last()
+    }
+}
+```
+
+Correcty, the compiler rejects empty tuples while tuples of other sizes are accepted:
+
+```rust
+().last();       // does not compile: no method named `last` found for type `()` in the current scope
+(1,).last();     // returns 1
+(1, "two").last(); // returns "two"
+etc.
+```
+
+The last example demonstrates how every second element of a tuple can be removed:
+
+```rust
+trait Halve {
+    type Output;
+
+    fn halve(self) -> Self::Output;
+}
+
+impl Halve for () {
+    type Output = ();
+
+    fn halve(self) {}
+}
+
+impl<ELEM1, ELEM2, TAIL: Halve> Halve for Tuple<ELEM1, Tuple<ELEM2, TAIL>> {
+    type Output = Tuple<ELEM1, TAIL>;
+
+    fn halve(self) -> Self::Output {
+        Tuple::new(self.elem, self.tail.tail)
+    }
+}
+```
+
+Results:
+
+```rust
+().halve();                   // returns ()
+(1,).halve()                  // does not compile: no method named `halve` found for type `(i32,)` in the current scope
+(1, "two").halve();           // returns (1,)
+(1, "two", 3.0, '4').halve(); // returns (1, 3.0)
+etc.
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The idea of this RFC to leave the tuple notation as is but to treat tuple type expressions as aliases according to the following pattern:
+
+```rust
+type () = (); // Not really an alias. Written down for completeness.
+type (A,) = Tuple<A, ()>;
+type (A, B) = Tuple<A, Tuple<B, ()>>;
+type (A, B, C) = Tuple<A, Tuple<B, Tuple<C, ()>>>;
+type (A, (B, C)) = Tuple<A, Tuple<Tuple<B, Tuple<C, ()>>, ()>>
+etc.
+```
+
+where `Tuple` is a new struct located in `std::ops` with the following definition:
+
+```rust
+struct Tuple<ELEM, TAIL> {
+    pub elem: ELEM,
+    pub tail: TAIL,
+}
+```
+
+This is everything needed to make the `Join` or `Last` traits from the guide section above work for tuples of any arity.
+
+## Required compiler changes
+
+- The compiler needs to treat any type `(ELEM, TAIL.0, ..., TAIL.n-1)` to be equivalent to `Tuple<ELEM, (TAIL.0, ..., TAIL.n-1)>`. This could work in the same way as `std::io::Result<T>` is considered equivalent to `core::result::Result<T, std::io::Error>`.
+- Equivalently, every tuple value `(elem, tail.0, ..., tail.n-1)` must be considered structurally equal to `Tuple { elem: elem, tail: (tail.0, ..., tail.n-1) }`.
+- Every tuple index access `tuple.n` must evaluate to `tuple{{.tail}^n}.elem`. In other words, `.tail` must be called `n` times before calling `.elem`.
```

Comment:
So essentially, you are moving tuples to the HIR lowering phases of the compiler and out of later phases. That turns tuples into essentially syntactic sugar. This is a nice simplification. On the other hand, this may also inflate compile times by giving the type checker later phases larger HIR trees to work with.

My overall sense is that it is hard to answer both the run-time and compile-time perf questions without implementing this in a PR and testing it out. Thus, if we are going to accept this, it would be best to run some experiments and get data to inform these questions.